rct229/schema/validate.py

```diff
@@ -26,6 +26,63 @@
 SCHEMA_OUTPUT_PATH = os.path.join(file_dir, SCHEMA_OUTPUT_KEY)
 
 
+def check_associated_list_length(rpd: dict) -> list[str]:
+    """
+    Check the length of the lists associated with the given jsonpaths.
+    Parameters
+    ----------
+    rpd
+
+    Returns list of object ids where the associated lists are not the same length
+    -------
+
+    """
+    mismatch_errors = []
+
+    associated_list_data_elements = [
+        ("use_units", "use"),
+        ("efficiency_metric_types", "efficiency_metric_values"),
+    ]
+    associated_list_jsonpaths = [
+        "$.ruleset_model_descriptions[*].buildings[*].building_segments[*].service_water_heating_uses[*]",
+        "$.ruleset_model_descriptions[*].buildings[*].building_segments[*].zones[*].spaces[*].service_water_heating_uses[*]",
+        "$.ruleset_model_descriptions[*].buildings[*].building_segments[*].heating_ventilating_air_conditioning_systems[*].heating_system",
+        "$.ruleset_model_descriptions[*].buildings[*].building_segments[*].heating_ventilating_air_conditioning_systems[*].preheat_system",
+        "$.ruleset_model_descriptions[*].buildings[*].building_segments[*].heating_ventilating_air_conditioning_systems[*].cooling_system",
+        "$.ruleset_model_descriptions[*].boilers[*]",
+        "$.ruleset_model_descriptions[*].chillers[*]",
+        "$.ruleset_model_descriptions[*].service_water_heating_equipment[*]",
+    ]
+
+    for jsonpath in associated_list_jsonpaths:
+        object_list = find_all(jsonpath, rpd)
+
+        for obj in object_list:
+            for key_1, key_2 in associated_list_data_elements:
+                has_key_1 = key_1 in obj
+                has_key_2 = key_2 in obj
+
+                if has_key_1 or has_key_2:
+                    if not (has_key_1 and has_key_2):  # One key is missing
+                        mismatch_errors.append(
+                            f"'{obj['id']}' has populated '{key_1 if has_key_1 else key_2}' but is missing '{key_1 if not has_key_1 else key_2}'."
+                        )
+                        continue
+
+                    # Ensure when both are lists they have the same length
+                    val_1, val_2 = obj[key_1], obj[key_2]
+                    if (
+                        isinstance(val_1, list)
+                        and isinstance(val_2, list)
+                        and len(val_1) != len(val_2)
+                    ):
```

Comment:
If schema validation has run before this function, then here you can simply do `val_1 and val_2 and len(val_1) != len(val_2)`
But I do also like to check for list to specify these values shall be list.